File: ios/RecaptchaEnterprise.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(RecaptchaEnterprise, NSObject)

RCT_EXTERN_METHOD(initializeRecaptcha:(NSString*)siteKey
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(executeAction:(NSString*)actionName
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(canUseRecaptcha:
                 withResolver:(RCTPromiseResolveBlock)resolve
                 withRejecter:(RCTPromiseRejectBlock)reject)


- (dispatch_queue_t)methodQueue
{
    return dispatch_get_main_queue();
}

+ (BOOL)requiresMainQueueSetup
{
  return YES;
}

@end

File: ios/RecaptchaEnterprise.swift
import RecaptchaEnterprise

@objc(RecaptchaEnterprise)
class RecaptchaEnterprise: NSObject {
    var recaptchaClient: RecaptchaClient?

    @objc(initializeRecaptcha:withResolver:withRejecter:)
    func initializeRecaptcha(siteKey: String, resolve: @escaping RCTPromiseResolveBlock,reject: @escaping RCTPromiseRejectBlock) -> Void {
        Recaptcha.getClient(siteKey: siteKey) { recaptchaClient, error in
            if let recaptchaClient = recaptchaClient {
              self.recaptchaClient = recaptchaClient
                resolve(nil)
            }
            if let error = error {
              print("RecaptchaClient creation error: \(error).")
                reject(String(error.errorCode), error.errorMessage, error)
            }
        }
    }

    @objc(executeAction:withResolver:withRejecter:)
    func executeAction(actionName: String, resolve: @escaping RCTPromiseResolveBlock,reject: @escaping RCTPromiseRejectBlock) -> Void {
        guard let recaptchaClient = self.recaptchaClient else {
            reject("NotInitializedClient", "RecaptchaClient is not initialized", nil)
            return
        }

        let actualAction = actionName.uppercased() == "LOGIN"
            ? RecaptchaAction.init(action: .login)
            : RecaptchaAction.init(customAction: actionName)

        recaptchaClient.execute(actualAction) { executeResult, error in
            if let executeResult = executeResult {
                resolve(executeResult.recaptchaToken)
            } else if let error = error {
                reject(String(error.errorCode), error.errorMessage, error)
            }
        }
    }

    @objc(canUseRecaptcha:withRejecter:)
    func canUseRecaptcha(resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
        resolve(true)
    }
}
